File: computing/interview/data-structures/cpp/queue.cpp
// Queue & Deque — std::queue (FIFO) and std::deque (double-ended)
#include <deque>
#include <iostream>
#include <queue>
#include <string>
#include <vector>

// Classic use-case: BFS on a simple graph
void bfs(int start, const std::vector<std::vector<int>>& adj) {
    std::vector<bool> visited(adj.size(), false);
    std::queue<int> q;
    q.push(start);
    visited[start] = true;

    std::cout << "BFS order: ";
    while (!q.empty()) {
        int node = q.front(); q.pop();   // O(1)
        std::cout << node << " ";
        for (int neighbor : adj[node]) {
            if (!visited[neighbor]) {
                visited[neighbor] = true;
                q.push(neighbor);        // O(1)
            }
        }
    }
    std::cout << "\n";
}

int main() {
    // --- Basic queue ---
    std::queue<int> q;
    q.push(10);   // enqueue — O(1)
    q.push(20);
    q.push(30);

    std::cout << "Front: " << q.front() << ", Back: " << q.back() << "\n";
    q.pop();   // dequeue — O(1)
    std::cout << "After dequeue, front: " << q.front() << "\n";

    // --- Deque (double-ended queue) ---
    std::deque<int> dq = {5, 6, 7};
    dq.push_front(4);   // O(1)
    dq.push_back(8);    // O(1)
    dq.pop_front();     // O(1)

    std::cout << "Deque: ";
    for (int x : dq) std::cout << x << " ";
    std::cout << "\n";

    // --- BFS demo ---
    //  0 -- 1 -- 3
    //  |    |
    //  2    4
    std::vector<std::vector<int>> adj = {
        {1, 2},    // 0
        {0, 3, 4}, // 1
        {0},       // 2
        {1},       // 3
        {1}        // 4
    };
    bfs(0, adj);

    return 0;
}

File: computing/interview/data-structures/python/queue.py
# Queue — collections.deque (O(1) at both ends) and queue.Queue (thread-safe)
from collections import deque
from queue import Queue


def bfs(start: int, adj: list[list[int]]) -> list[int]:
    """BFS traversal — returns visited order."""
    visited = [False] * len(adj)
    order = []
    q = deque([start])
    visited[start] = True
    while q:
        node = q.popleft()          # O(1)
        order.append(node)
        for neighbor in adj[node]:
            if not visited[neighbor]:
                visited[neighbor] = True
                q.append(neighbor)  # O(1)
    return order


# --- deque as FIFO queue ---
q = deque()
q.append(10)    # enqueue — O(1)
q.append(20)
q.append(30)
print("Front:", q[0])
print("Dequeued:", q.popleft())   # O(1)
print("Queue:", list(q))

# --- queue.Queue: thread-safe FIFO ---
tq: Queue[str] = Queue()
tq.put("task-A")
tq.put("task-B")
tq.put("task-C")
print("Thread-safe queue size:", tq.qsize())
print("Got:", tq.get())   # blocks if empty

# --- BFS demo ---
#  0 -- 1 -- 3
#  |    |
#  2    4
adj = [[1, 2], [0, 3, 4], [0], [1], [1]]
print("BFS from 0:", bfs(0, adj))
